File: latest/week2/workshop6_m1/W06_M01_demo.R
# ============================================================
# Workshop 6 — Module 1: Time as a Visual Dimension
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(lubridate); dir.create("output", showWarnings = FALSE)

# ── 1. SIMULATED DAILY TIME SERIES ──────────────────────
set.seed(42)
dates <- seq(as.Date("2020-01-01"), as.Date("2023-12-31"), by = "day")
n <- length(dates)
trend <- 100 + 0.05 * (1:n)
seasonal <- 15 * sin(2 * pi * (1:n) / 365)
residual <- rnorm(n, 0, 5)
value <- trend + seasonal + residual

daily <- tibble(date = dates, value = value)
cat("Daily series:", nrow(daily), "observations\n")
cat("Range:", format(min(daily$date)), "to", format(max(daily$date)), "\n")

# ── 2. GRANULARITY COMPARISON ────────────────────────────
weekly  <- daily |> mutate(week  = floor_date(date, "week"))  |> group_by(week)  |> summarise(value = mean(value)) |> rename(date = week)
monthly <- daily |> mutate(month = floor_date(date, "month")) |> group_by(month) |> summarise(value = mean(value)) |> rename(date = month)
yearly  <- daily |> mutate(year  = floor_date(date, "year"))  |> group_by(year)  |> summarise(value = mean(value)) |> rename(date = year)

p_daily   <- ggplot(daily, aes(x = date, y = value)) + geom_line(color = "#BBBBBB", linewidth = 0.3) +
  theme_minimal(base_size = 8) + labs(title = "(a) Daily")
p_weekly  <- ggplot(weekly, aes(x = date, y = value)) + geom_line(color = "#1565C0", linewidth = 0.8) +
  theme_minimal(base_size = 8) + labs(title = "(b) Weekly")
p_monthly <- ggplot(monthly, aes(x = date, y = value)) + geom_line(color = "#E53935", linewidth = 1) + geom_point(color = "#E53935", size = 1.5) +
  theme_minimal(base_size = 8) + labs(title = "(c) Monthly")
p_yearly  <- ggplot(yearly, aes(x = date, y = value)) + geom_line(color = "#2E7D32", linewidth = 1.5) + geom_point(color = "#2E7D32", size = 3) +
  theme_minimal(base_size = 8) + labs(title = "(d) Yearly")

library(patchwork)
p_gran <- (p_daily | p_weekly) / (p_monthly | p_yearly) +
  plot_annotation(title = "Same Data, Four Granularities", tag_levels = "a")
ggsave("output/granularity.png", p_gran, width = 12, height = 7, dpi = 300)

# ── 3. DECOMPOSITION ────────────────────────────────────
# Convert to ts object (frequency = 365 for daily data)
ts_data <- ts(daily$value, frequency = 365)

# STL decomposition (robust)
stl_fit <- stl(ts_data, s.window = "periodic")

# Plot the four panels
png("output/decomposition.png", width = 10, height = 8, units = "in", res = 300)
plot(stl_fit, main = "STL Decomposition: Trend + Seasonal + Residual")
dev.off()

# Extract components
components <- tibble(
  date = daily$date,
  observed = daily$value,
  trend = as.numeric(stl_fit$time.series[, "trend"]),
  seasonal = as.numeric(stl_fit$time.series[, "seasonal"]),
  remainder = as.numeric(stl_fit$time.series[, "remainder"]))

# ggplot version
p_decomp <- components |>
  pivot_longer(-date, names_to = "component", values_to = "val") |>
  mutate(component = factor(component, levels = c("observed", "trend", "seasonal", "remainder"))) |>
  ggplot(aes(x = date, y = val)) +
  geom_line(color = "#1565C0", linewidth = 0.3) +
  facet_wrap(~component, ncol = 1, scales = "free_y") +
  theme_minimal(base_size = 8) +
  labs(title = "STL Decomposition (ggplot version)", y = NULL)
ggsave("output/decomposition_gg.png", p_decomp, width = 10, height = 8, dpi = 300)

# ── 4. CHART TYPE COMPARISON ────────────────────────────
# Line
p_line <- ggplot(monthly, aes(x = date, y = value)) +
  geom_line(color = "#1565C0", linewidth = 1) +
  theme_minimal(base_size = 8) + labs(title = "(a) Line Chart")

# Area
p_area <- ggplot(monthly, aes(x = date, y = value)) +
  geom_area(fill = "#1565C0", alpha = 0.3, color = "#1565C0", linewidth = 0.5) +
  theme_minimal(base_size = 8) + labs(title = "(b) Area Chart")

# Step
p_step <- ggplot(monthly, aes(x = date, y = value)) +
  geom_step(color = "#E53935", linewidth = 1) +
  theme_minimal(base_size = 8) + labs(title = "(c) Step Chart")

# Bar
p_bar <- ggplot(monthly, aes(x = date, y = value)) +
  geom_col(fill = "#7B1FA2", width = 20, alpha = 0.7) +
  theme_minimal(base_size = 8) + labs(title = "(d) Bar Chart (monthly)")

p_types <- (p_line | p_area) / (p_step | p_bar) +
  plot_annotation(title = "Four Temporal Chart Types: Same Monthly Data")
ggsave("output/chart_types.png", p_types, width = 12, height = 7, dpi = 300)

# ── 5. PITFALL: CATEGORICAL vs DATE AXIS ────────────────
# Bad: dates as factor (even spacing)
bad_dates <- c("2020-01", "2020-06", "2020-12", "2022-01", "2023-06")
bad_vals <- c(10, 15, 18, 22, 28)
p_bad <- tibble(date = bad_dates, val = bad_vals) |>
  ggplot(aes(x = date, y = val, group = 1)) +
  geom_line(color = "#E53935", linewidth = 1) + geom_point(color = "#E53935", size = 3) +
  theme_minimal(base_size = 9) +
  labs(title = "BAD: Dates as categories (even spacing)")

# Good: proper date axis
p_good <- tibble(date = ymd(paste0(bad_dates, "-01")), val = bad_vals) |>
  ggplot(aes(x = date, y = val)) +
  geom_line(color = "#1565C0", linewidth = 1) + geom_point(color = "#1565C0", size = 3) +
  scale_x_date(date_breaks = "6 months", date_labels = "%Y-%m") +
  theme_minimal(base_size = 9) +
  labs(title = "GOOD: True date axis (real spacing)")

ggsave("output/pitfall_axis.png", p_bad | p_good, width = 12, height = 4, dpi = 300)

cat("\n── All W06-M01 R outputs saved ──\n")

File: latest/week2/workshop6_m1/W06_M01_demo.py
"""W06-M01: Time as a Visual Dimension — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.dates import DateFormatter, MonthLocator, YearLocator
os.makedirs("output", exist_ok=True); np.random.seed(42)

# 1. SIMULATED DAILY TIME SERIES
dates = pd.date_range("2020-01-01", "2023-12-31", freq="D")
n = len(dates)
trend = 100 + 0.05 * np.arange(n)
seasonal = 15 * np.sin(2 * np.pi * np.arange(n) / 365)
residual = np.random.normal(0, 5, n)
value = trend + seasonal + residual
daily = pd.DataFrame({"date": dates, "value": value}).set_index("date")
print(f"Daily series: {len(daily)} observations, {daily.index.min()} to {daily.index.max()}")

# 2. GRANULARITY COMPARISON
weekly = daily.resample("W").mean()
monthly = daily.resample("ME").mean()
yearly = daily.resample("YE").mean()

fig, axes = plt.subplots(2, 2, figsize=(12, 7))
for ax, data, title, color in zip(axes.flatten(),
    [daily, weekly, monthly, yearly],
    ["(a) Daily", "(b) Weekly", "(c) Monthly", "(d) Yearly"],
    ["#BBBBBB", "#1565C0", "#E53935", "#2E7D32"]):
    ax.plot(data.index, data["value"], color=color, lw=0.5 if "Daily" in title else 1.5)
    if "Monthly" in title or "Yearly" in title:
        ax.plot(data.index, data["value"], "o", color=color, markersize=3 if "Monthly" in title else 6)
    ax.set_title(title, fontsize=9, fontweight="bold", color=color)
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False); ax.tick_params(labelsize=6)
fig.suptitle("Same Data, Four Granularities", fontsize=12, fontweight="bold")
plt.tight_layout(); plt.savefig("output/granularity.png", dpi=300); plt.close()

# 3. DECOMPOSITION
try:
    from statsmodels.tsa.seasonal import STL
    stl = STL(daily["value"], period=365, robust=True)
    res = stl.fit()

    fig, axes = plt.subplots(4, 1, figsize=(10, 8), sharex=True)
    panels = [(daily["value"], "Observed", "#333"),
              (res.trend, "Trend", "#1565C0"),
              (res.seasonal, "Seasonal", "#2E7D32"),
              (res.resid, "Residual", "#E53935")]
    for ax, (data, title, color) in zip(axes, panels):
        ax.plot(data.index, data.values, color=color, lw=0.5)
        ax.set_title(title, fontsize=9, fontweight="bold", color=color, loc="left")
        ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False); ax.tick_params(labelsize=6)
    fig.suptitle("STL Decomposition: Trend + Seasonal + Residual", fontsize=12, fontweight="bold")
    plt.tight_layout(); plt.savefig("output/decomposition.png", dpi=300); plt.close()
    print("STL decomposition saved")
except ImportError:
    print("statsmodels not available — skipping decomposition")

# 4. CHART TYPE COMPARISON
fig, axes = plt.subplots(2, 2, figsize=(12, 7))
# Line
axes[0,0].plot(monthly.index, monthly["value"], color="#1565C0", lw=1.5)
axes[0,0].set_title("(a) Line Chart", fontsize=9, fontweight="bold")
# Area
axes[0,1].fill_between(monthly.index, monthly["value"], alpha=0.3, color="#1565C0")
axes[0,1].plot(monthly.index, monthly["value"], color="#1565C0", lw=0.8)
axes[0,1].set_title("(b) Area Chart", fontsize=9, fontweight="bold")
# Step
axes[1,0].step(monthly.index, monthly["value"], color="#E53935", lw=1.5, where="mid")
axes[1,0].set_title("(c) Step Chart", fontsize=9, fontweight="bold")
# Bar
axes[1,1].bar(monthly.index, monthly["value"], width=25, color="#7B1FA2", alpha=0.7)
axes[1,1].set_title("(d) Bar (monthly)", fontsize=9, fontweight="bold")
for ax in axes.flatten():
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
    ax.tick_params(labelsize=6); ax.xaxis.set_major_formatter(DateFormatter("%Y-%m"))
fig.suptitle("Four Temporal Chart Types: Same Monthly Data", fontsize=12, fontweight="bold")
plt.tight_layout(); plt.savefig("output/chart_types.png", dpi=300); plt.close()

# 5. PITFALL: CATEGORICAL vs DATE AXIS
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 4))
bad_dates_str = ["2020-01", "2020-06", "2020-12", "2022-01", "2023-06"]
vals = [10, 15, 18, 22, 28]
ax1.plot(range(len(bad_dates_str)), vals, "-o", color="#E53935", lw=2, markersize=6)
ax1.set_xticks(range(len(bad_dates_str))); ax1.set_xticklabels(bad_dates_str, fontsize=7, rotation=30)
ax1.set_title("BAD: Categorical spacing", fontsize=9, fontweight="bold", color="#C62828")

real_dates = pd.to_datetime([d + "-01" for d in bad_dates_str])
ax2.plot(real_dates, vals, "-o", color="#1565C0", lw=2, markersize=6)
ax2.set_title("GOOD: True date axis", fontsize=9, fontweight="bold", color="#1565C0")
ax2.xaxis.set_major_formatter(DateFormatter("%Y-%m"))
for ax in [ax1, ax2]: ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Time Axis Pitfall: Always Use True Date Spacing", fontsize=11, fontweight="bold")
plt.tight_layout(); plt.savefig("output/pitfall_axis.png", dpi=300); plt.close()

# 6. DATE FORMATTING CHEATSHEET
print("\n── pandas resample codes ──")
print("  D=daily, W=weekly, ME=month-end, MS=month-start")
print("  QE=quarter-end, YE=year-end")
print("── matplotlib DateFormatter ──")
print("  %Y=year, %m=month, %d=day, %b=abbrev month, %B=full month")

print("\nAll W06-M01 Python outputs saved")
